File: asion-account/asion-account-web/src/main/java/org/asion/account/web/AccountController.java
package org.asion.account.web;

import org.asion.account.domain.model.Account;
import org.asion.account.domain.application.AccountManager;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import javax.validation.Valid;

/**
 * user controller.
 * @author Asion.
 * @since 16/5/29.
 */
@Controller
@RequestMapping("/user/")
public class AccountController {

    @Autowired
    private AccountManager accountManager;

    @RequestMapping("list")
    public ModelAndView list() {
        Iterable<Account> accounts = accountManager.findAll();
        return new ModelAndView("user/list", "users", accounts);
    }

    @RequestMapping("{id}")
    public ModelAndView view(@PathVariable("id") Long id) {
        Account account = accountManager.findOne(id);
        return new ModelAndView("user/view", "user", account);
    }

    @RequestMapping(params = "form", method = RequestMethod.GET)
    public String createForm(@ModelAttribute Account account) {
        return "user/form";
    }

    @RequestMapping(value = "create",method = RequestMethod.POST)
    public ModelAndView create(@Valid Account account, BindingResult result, RedirectAttributes redirect) {
        if (result.hasErrors()) {
            return new ModelAndView("user/form", "formErrors", result.getAllErrors());
        }
        account = accountManager.save(account);
        redirect.addFlashAttribute("globaluser", "Successfully save a new user");
        return new ModelAndView("redirect:/user/{user.id}", "user.id", account.getId());
    }

    @RequestMapping("foo")
    public String foo() {
        throw new RuntimeException("Expected exception in controller");
    }

    @RequestMapping(value = "delete/{id}")
    public ModelAndView delete(@PathVariable("id") Long id) {
        accountManager.delete(id);
        Iterable<Account> accounts = accountManager.findAll();
        return new ModelAndView("user/list", "users", accounts);
    }

    @RequestMapping(value = "modify/{id}", method = RequestMethod.GET)
    public ModelAndView modifyForm(@PathVariable("id") Long id) {
        Account account = accountManager.findOne(id);
        return new ModelAndView("user/form", "user", account);
    }

}

File: asion-account/asion-account-web/src/main/java/io/vincent/account/web/AccountController.kt
package io.vincent.account.web

import com.alibaba.dubbo.config.annotation.Reference
import io.vincent.account.domain.application.AccountManager
import io.vincent.account.domain.model.Account
import org.springframework.stereotype.Controller
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.mvc.support.RedirectAttributes
import java.util.*
import javax.validation.Valid

/**
 * user controller.
 *
 * @author Asion.
 * @since 16/5/29.
 */
@Controller
@RequestMapping("/account")
class AccountController {

    @Reference(version = "1.0.0", application = "\${dubbo.application.id}", url = "dubbo://dubbo.asion.org:2008")
    private val accountManager: AccountManager? = null

    @RequestMapping("/list")
    fun list(): ModelAndView {
        val accounts = accountManager!!.findPage(1, 10)
        val list = accounts.list
        return ModelAndView("account/list", "accounts", list)
    }

    @RequestMapping("/{id}")
    fun view(@PathVariable("id") id: Long?): ModelAndView {
        val account = accountManager!!.findOne(id!!)
        return ModelAndView("account/view", "account", account!!)
    }

    @RequestMapping(value = "/", params = arrayOf("form"), method = arrayOf(RequestMethod.GET))
    fun createForm(@ModelAttribute account: Account): String {
        return "account/form"
    }

    @RequestMapping(value = "/create", method = arrayOf(RequestMethod.POST))
    fun create(@Valid account: Account, result: BindingResult, redirect: RedirectAttributes): ModelAndView {
        var account = account
        if (result.hasErrors()) {
            return ModelAndView("account/form", "formErrors", result.allErrors)
        }
        account = accountManager!!.create(account)
        redirect.addFlashAttribute("globalAccount", "Successfully save a new user")
        return ModelAndView("redirect:/account/{account.id}", "account.id", account.id!!)
    }

    @RequestMapping("/foo")
    fun foo(): String {
        throw RuntimeException("Expected exception in controller")
    }

    @RequestMapping(value = "/delete/{id}")
    fun delete(@PathVariable("id") id: Long?): ModelAndView {
        //        accountManager.delete(id);
        val accounts = ArrayList<Account>()
        return ModelAndView("account/list", "accounts", accounts)
    }

    @RequestMapping(value = "/modify/{id}", method = arrayOf(RequestMethod.GET))
    fun modifyForm(@PathVariable("id") id: Long?): ModelAndView {
        val account = accountManager!!.findOne(id!!)
        return ModelAndView("account/form", "account", account!!)
    }

}
